### DISCUSSION TITLE
[Immunization] How to close Tasks based on the selected items on the Record All forms | Code or Content Changes

### DISCUSSION THREAD

**Discussion by dubdabasoduba** [opensrp/fhircore#2143] [Ideas]
- We have 2 options to achieve this.

1. **Content Changes**
    - The `Questionnaire.item.answerOption.valueCoding.code` is the `Task.id` for the Tasks represented by
      the `Questionnaire.item.answerOption`.
        - Write a structure map function to update `Task.status` to `completed`.
        - Use the `Questionnaire.item.answerOption.valueCoding.code` as the `Task.id.value`.

- **Advantages**
    - We don't have to add new code to process the Tasks status.
- **Disadvantages**
    - This behaves very differently from the Tasks closure for a form launched by specific tasks. This increases the
      maintainability complexity
    - The tasking workflow is very key for the FHIR Core functionality. Letting the content authors determine when the
      flow is complete is a risk.

2. **Code Changes**
    - Update the ActionParameters to add a new parameter to define the Questionnaire Item that holds displays the
      CheckBox showing the Tasks.
        - This parameter should be a string containing the `Questionnaire.item.linkId`
    - Write a function that iterates through the QuestionnaireResponse to fetch the answers with
      a `QuestionnaireResponse.item.linkId` that matches the definition on the ActionParameters.
        - Get the `QuestionnaireResponse.item.answer.valueCoding.code` from each answer.
        - Use the list of `QuestionnaireResponse.item.answer.valueCoding.code` as Task IDs to fetch Tasks from the
          SQLite database.
        - Update all the returned `Task.status` to `completed`

- **Advantages**
    - We maintain the same architecture where the code handles closing the loop on the tasking workflow.
    - The users don't have to worry about determining when tasks are completed.
- **Disadvantages**
    - We will be adding extra code targeting Tasks status updates.

**Comment by pld** [comment]
I prefer the content approach, at least at this point. It's hard to know yet what is going to be the best way to handle tasking workflows, if there was an obvious general approach, or one emerges, then I'd agree we should consider hard-coding that, but I don't think we have the data to determine that yet
